lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs

```diff
@@ -108,158 +78,86 @@ internal sealed class UpdateTaskHost<TMapInput, TMapOutput, TResult> : ITask, IO
+                        MapControlMessage.AnotherRound))
                 {
-                    Logger.Log(Level.Error,
-                        "Received AggregateException. The cancellation token is: {0}.",
-                        _cancellationSource.IsCancellationRequested);
-                    if (e.InnerException != null)
-                    {
-                        Logger.Log(Level.Error,
-                            "InnerException {0}, with message {1}.",
-                            e.InnerException.GetType(),
-                            e.InnerException.Message);
-                    }
-                    if (!_cancellationSource.IsCancellationRequested)
-                    {
-                        if (e.InnerException != null && e.InnerException is IOException)
-                        {
-                            Logger.Log(Level.Error,
-                                "UpdateTaskHost is throwing IMRUTaskGroupCommunicationException with cancellation token: {0}.",
-                                _cancellationSource.IsCancellationRequested);
-                            throw new IMRUTaskGroupCommunicationException(TaskManager.TaskGroupCommunicationError);
-                        }
-                        else
-                        {
-                            throw e;
-                        }
-                    }
+                    _dataAndControlMessageSender.Send(message);
                 }
-                else
+
+                var input = _dataReceiver.Reduce(_cancellationSource);
+
+                if (_invokeGc)
                 {
-                    Logger.Log(Level.Error,
-                       "UpdateTaskHost is throwing Excetion {0}, messge {1} with cancellation token: {2} and StackTrace {3}.",
-                       e.GetType(),
-                       e.Message,
-                       _cancellationSource.IsCancellationRequested,
-                       e.StackTrace);
-                    if (!_cancellationSource.IsCancellationRequested)
-                    {
-                        throw e;
-                    }
+                    Logger.Log(Level.Verbose, "Calling Garbage Collector");
+                    GC.Collect();
+                    GC.WaitForPendingFinalizers();
                 }
-            }
-            finally
-            {
+
                 try
                 {
-                    _resultHandler.Dispose();
+                    updateResult = _updateTask.Update(input);
                 }
                 catch (Exception e)
                 {
-                    Logger.Log(Level.Error, "Exception in dispose result handler.", e);
-                    //// TODO throw proper exceptions JIRA REEF-1492
+                    HandleTaskAppException(e);
+                }
+
+                if (updateResult.HasResult)
+                {
+                    _resultHandler.HandleResult(updateResult.Result);
```

Comment:
Why do we wrap `updateResult = _updateTask.Initialize();` in app exception but don't wrap `HandleResult` in app exception? Isn't `HandleResult` also provided by client code?